lib/src/functions/map.dart

```diff
@@ -73,6 +91,26 @@ final _hasKey = _function("has-key", r"$map, $key", (arguments) {
   return SassBoolean(map.contents.containsKey(key));
 });
 
+SassMap _set(SassMap map, List<Value> keys, Value value) {
+  var mutableMap = Map.of(map.contents);
+  var key = keys.first;
+
+  if (keys.length == 1) {
+    mutableMap[key] = value;
+    return SassMap(mutableMap);
+  }
+
+  SassMap nestedMap;
+  try {
+    nestedMap = mutableMap[key].assertMap();
+  } on SassScriptException {
+    nestedMap = SassMap({});
+  }
+
+  mutableMap[key] = _set(nestedMap, keys.sublist(1), value);
```

Comment:
This is pretty nitty, but creating a separate sublist for every recursion could result in a lot of extra data structure allocations which are relatively expensive. It may be worth writing this in an iterative style instead of recursive to avoid that.

Reply:
I had trouble expressing this nicely as iterative, so instead I passed an index into the array to avoid sublisting it over and over. WDYT?

Reply:
I can see why a directly iterative solution would be messy, since you'd need to track an explicit stack of parent maps *and* their associated keys. I think this solution works well enough.